src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
+        // before the report in which the keycodes they are meant to
+        // modify gets sent. Instead, we use this variable, which will
+        // get bit-shifted each cycle, and trigger the release when it
+        // gets to 1 (which is why it's written as a bit-shifted 1 here).
+        release_countdown_ = 1 << 1;
       }
     }
-
-    return EventHandlerResult::EVENT_CONSUMED;
+    return EventHandlerResult::OK;
   }
 
-  // ordinary key here, with some event
+  // We know that `mapped_key` is a OneShot key, and the call to
+  // `isOneShotKey_()` above ensures that the index is not out of
+  // bounds, so this is now valid:
+  uint8_t n = mapped_key.getRaw() - ranges::OS_FIRST;
+
+  if (keyToggledOn(key_state)) {
+    state_[n].pressed = true;
+    if (! state_[n].active) {
+      // This is the first press of an inactive OneShot key, so we
+      // simply activate it, setting `active` to true.
+      // `activateOneShot()` sends the toggled on event, and in
+      // subsequent cycles, the `beforeReportingState()` hook will
+      // send `sustainOneShot()`, sending the hold "event".
+      activateOneShot(n);
+    } else if (state_[n].sticky) {
+      // If the OneShot key is already sticky, and it gets pressed, we
+      // remove its `active` state flag. As long as it is still held
+      // (`pressed == true`), it will continue to be translated into
+      // its corresponding modifier in `beforeReportingState()`.
+      state_[n].active = false;
+      state_[n].sticky = false;
+    } else {
+      // The OneShot key has been pressed while it is active, but not
+      // (yet) sticky. First we determine if it's a candidate for
+      // becoming sticky: it can only become sticky if the same key is
+      // pressed twice in a row, and that key is configured to be
+      // sticky.
+      if ((key_addr == prev_key_addr_) && state_[n].stickable) {
+        // In addition, it must have been "double-tapped": pressed
+        // twice in less than `double_tap_time_out` milliseconds. But
+        // if that variable isn't set (negative), we use the default
+        // `time_out` value instead.
+        uint16_t dtto = (double_tap_time_out < 0) ? time_out : double_tap_time_out;
+        if (! Runtime.hasTimeExpired(start_time_, dtto)) {
+          // Making a OneShot sticky is as simple as setting its
+          // `sticky` flag; the other hook functions handle the rest.
+          state_[n].sticky = true;
+        } else {
+          // If two presses weren't quick enough, this second press
+          // will terminate the OneShot behaviour (see above).
+          state_[n].active = false;
+        }
+      } else {
+        // If this OneShot key is not "stickable", we deactivate
+        // it. It still remains in effect until it is released.
+        state_[n].active = false;
+      }
+    }
+    prev_key_addr_ = key_addr;
+
+  } else if (keyToggledOff(key_state)) {
+    state_[n].pressed = false;
```

Comment:
I'm now guaranteeing that `pressed` will be true when a OneShot key is pressed, and false when it's not…unless there are two of the same OneShot key pressed at the same time, in which case it will get unset when the first one gets released. This is an unfortunate side effect of tracking `Key` values instead of `KeyAddr` values. The problem could be fixed by searching through the whole keymap for held OneShot keys any time one is released, but I don't think it's worth the trouble.